### PR TITLE
[pybind] Changed union set_* to ensure_*, added equality operators, misc improvements

### PR DESCRIPTION
**Pull request by michaelahn-figure**
### Changes
- Adds `operator==` and `operator!=` to structs and object API tables
- Fixed `bool` fields appearing as python `int` type in method arguments instead of `bool`.
- Fixed struct array fields not being writable
- Mutating a packed flatbuffer scalar field now throws an error if unsuccessful (i.e. there is no memory backing the field)
  - In c++, the `mutate_<field>` method returns `false` in such a case
- Ensure no code generation for vector of union fields -- support for this will be added in a future PR
- Changed `set_<field>` to `ensure_<field>` for union fields; see below for discussion.

A unit test in project-x tests this behavior; not included in this repo as we don't have any CI setup for it, and there is no standard devcontainer setup to run it locally. I'll send this out in a separate internal PR.

#### `ensure_<field>` for unions

Previously, we generated a `set_<field_name>(<UnionEnumType>)` to internally allocate a union field (as opposed to the regular way of copying a subtype instance through assignment). This method was somewhat cumbersome:
- This is the only place in the generated code that uses the union enum values. The user needs to understand which enum values correspond to which message types.
- This always resets the field to a default-initialized sub type.

This changes the method to be `ensure_<field_name>(subtype)`. This has a similar purpose as `ensure_<field_name>()` for non-union optional fields, except the specific subtype must be specified. This has the behavior of:
1. If the field is unset, allocates a new instance of `subtype`.
2. If the field was already set, throws an error if the existing subtype does not match `subtype`.
3. Returns the instance.

- The python type object is converted to the underlying union enum under the hood; the user does not need to use or know about the union enum values.

```python
a = FooT()
a.b  # type is BarT | BazT | None
b = a.ensure_b(BazT)  # type is always BazT
```

Caveat: pybind currently doesn't provide the information to type systems that `ensure_b(T)` constrains the return type to `T`. Most type checkers will think `ensure_b(T) -> BarT | BazT | None`, so an additional `cast` or type assertion may be necessary to appease the type check.

Example generated code looks like:
```c++
  monster_t.def(
      "ensure_equipped",
      [](MyGame::Sample::MonsterT &self, py::type union_type) -> std::optional<std::variant<MyGame::Sample::WeaponT*>> {
        MyGame::Sample::Equipment union_enum;
        if (union_type.is(py::type::of<MyGame::Sample::WeaponT>())) {
          union_enum = MyGame::Sample::Equipment_Weapon;
        }
        else {
          throw py::value_error(py::str("{} is not part of union Equipment").format(union_type));
        }
        bool holds_none = self.equipped.type == MyGame::Sample::Equipment_NONE;
        if (!holds_none && self.equipped.type != union_enum) {
          throw py::value_error(py::str("Union field is already set to {}").format(self.equipped.type));
        }
        switch (union_enum) {
          case MyGame::Sample::Equipment_NONE:
            return std::nullopt;
          case MyGame::Sample::Equipment_Weapon:
            if (holds_none) self.equipped.Set(MyGame::Sample::WeaponT());
            return self.equipped.AsWeapon();
          default:
            throw std::runtime_error("Invalid variant type");
        }
      }, py::return_value_policy::reference_internal);
```

### CODE REVIEW

File: src/idl_gen_pybind.cpp [hunk 1]
```diff
@@ -876,8 +922,8 @@ class PybindGenerator : public BaseGenerator {
         code_ += Indent(3) +
                  "[](const {{CPP_TYPE}} &self) -> "
                  "std::optional<{{CPP_FIELD_TYPE}}*> {";
-        code_ +=
-            Indent(4) + "if (self.{{CPP_FIELD}} == nullptr) { return std::nullopt; }";
```
Comment (eneev-a) [inline src/idl_gen_pybind.cpp:880]: are formatting changes intended?

Comment (michaelahn-figure) [reply src/idl_gen_pybind.cpp:880]: Yeah, clang-format seems to have changed its mind on this. Haven't investigated why, but I'll see if these spurious formatting changes happen again.

File: src/idl_gen_pybind.cpp [hunk 2]
```diff
@@ -934,6 +980,12 @@ class PybindGenerator : public BaseGenerator {
     // The memoryview keeps the bytearray buffer alive.
     code_ += Indent() + "}, py::keep_alive<0, 2>());";
 
+    // operators.
+    if (opts_.gen_compare) {
+      code_ += Indent() + "{{BIND_VAR}}.def(py::self == py::self);";
```
Comment (eneev-a) [inline src/idl_gen_pybind.cpp:985]: :exploding_head:

I am thinking of the machinery behind the `.def(py::self == py::self)` statement and I hope that at figure we won't do use anything like this. DSLs can be amazing, but learning C++ itself is hard enough.

But I have to admit that this does look really neat.

**Review by eneev-a — APPROVED** [review]
:heavy_check_mark:

**Comment by e7c-figure** [comment]
> A unit test in project-x tests this behavior; not included in this repo as we don't have any CI setup for it, and there is no standard devcontainer setup to run it locally. I'll send this out in a separate internal PR.

+1 let's keep all of our tests in project-x if we can.

**Review by e7c-figure — APPROVED** [review]
LGTM

**Pull request merged**